### PR TITLE
Big faucet refactoring

### LINKED ISSUE #48: Add `COSMOS_ENABLED` CI test to faucet
Since we have the nice test setup for CI, let's add some tests to faucet to show it works with the proper balance. If you need to give more funds to a starting account (for the faucet), there are some instructions here: https://github.com/cosmwasm/wasmd/blob/master/README.md#ci

Just add a few more cosmos addresses after `setup.sh` and they will all have funds in the genesis. Then commit the new `template`

### LINKED ISSUE #65: Remove globalFractionalDigits from faucet
This whole code is a hack:

```ts
// Load this from connection?
let globalFractionalDigits: number | undefined;

export function setFractionalDigits(input: number): void {
  globalFractionalDigits = input;
}

export function getFractionalDigits(): number {
  if (globalFractionalDigits === undefined) {
    throw new Error("Fractional digits not set");
  }
  return globalFractionalDigits;
}
```

We should have a _ticker to digits_ function, which can easily be implemented in `codec.ts`, where the token configuration is available.

### PR DESCRIPTION
**Pull request by webmaster128**
Closes #48
Closes #65

All those free functions are now grouped into the classes `TokenManager`, `Faucet` and `Webserver`. Many tests for `Faucet` exist now, that connect to the local development chain.

### CODE REVIEW

**Comment by ethanfrey** [comment]
Wow. Major overhaul

**Comment by webmaster128** [comment]
I wonder where this CI fail comes from: https://circleci.com/gh/confio/workflows/cosmwasm-js/tree/no-global-state (the first run). This PR only changes content of the faucet package, which is not used in the deployment script.

**Comment by ethanfrey** [comment]
We get: `Error: Uploading failed with code: 111222; log: 'internal error'` which is useless... (something broke in the server).

For better errors, we need to add `--trace` to the `wasmd` command and use at least docker image `v0.0.3`. Note that in `v0.0.4` I fixed the client so it would properly report the messages with 500 errors (from queries).

**Comment by ethanfrey** [comment]
Rrr... the last advice requires updating the docker image.
Will do that now

**Comment by ethanfrey** [comment]
Okay, a `v0.0.5` docker image is building that has `--trace` enabled by default. I thought I did that already, but I guess only for local testing.

File: packages/faucet/src/faucet.spec.ts [hunk 1]
```diff
@@ -0,0 +1,204 @@
...
+        }),
+      ]);
+      expect(Number.parseInt(distributorBalance[0].quantity, 10)).toBeGreaterThanOrEqual(80_000000);
+      expect(Number.parseInt(distributorBalance[1].quantity, 10)).toBeGreaterThanOrEqual(80_000000);
+      connection.disconnect();
+    });
+  });
+
+  describe("credit", () => {
+    it("works for fee token", async () => {
```
Comment (ethanfrey) [inline packages/faucet/src/faucet.spec.ts]: nice set of tests

File: packages/faucet/src/faucet.ts [hunk 2]
```diff
@@ -1,37 +1,169 @@
...
+import { UserProfile } from "@iov/keycontrol";
+import { sleep } from "@iov/utils";
+
+import { identityToAddress } from "./addresses";
+import { debugAccount, logAccountsState, logSendJob } from "./debugging";
+import { availableTokensFromHolder, identitiesOfFirstWallet } from "./multichainhelpers";
+import { TokenManager } from "./tokenmanager";
+import { SendJob } from "./types";
+
+export class Faucet {
```
Comment (ethanfrey) [inline packages/faucet/src/faucet.ts]: Nice to extract this from those helpers :+1:

File: packages/faucet/src/faucet.spec.ts [hunk 3]
```diff
@@ -0,0 +1,204 @@
...
+    },
+    {
+      fractionalDigits: 6,
+      name: "Staking Token",
+      ticker: "STAKE",
+      denom: "ustake",
+    },
+  ],
+  erc20Tokens: [
+    // {
```
Comment (ethanfrey) [inline packages/faucet/src/faucet.spec.ts]: Cool stuff coming soon... :smile:

File: packages/faucet/src/faucet.spec.ts [hunk 4]
```diff
@@ -0,0 +1,204 @@
...
+}
+
+const httpUrl = "http://localhost:1317";
+const defaultConfig: TokenConfiguration = {
+  bankTokens: [
+    {
+      fractionalDigits: 6,
+      name: "Fee Token",
+      ticker: "COSM",
+      denom: "ucosm",
```
[the remark this answers is not in the document]
Comment (webmaster128) [reply packages/faucet/src/faucet.spec.ts]: Those will potientially diverge. The test runs against a local development chain. The config in constants can have the testnet values hardcoded.

File: packages/faucet/src/constants.ts [hunk 5]
```diff
@@ -1,4 +1,24 @@
...
 export const mnemonic: string | undefined = process.env.FAUCET_MNEMONIC;
+
+export const addressPrefix = "cosmos";
+export const tokenConfig: TokenConfiguration = {
+  bankTokens: [
+    {
+      fractionalDigits: 6,
+      name: "Fee Token",
+      ticker: "COSM",
+      denom: "cosm",
```
[the remark this answers is not in the document]
Comment (webmaster128) [reply packages/faucet/src/constants.ts]: Thanks. Done in 062b003a4a1843caf2475810a234ce6d9ffce769

**Pull request merged**